File: leetcode-cn/37.解数独.py
#
# @lc app=leetcode.cn id=37 lang=python3
#
# [37] 解数独
#

# @lc code=start
from typing import List


class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        line = [[False] * 9 for _ in range(9)]
        column = [[False] * 9 for _ in range(9)]
        block = [[[False] * 9 for _ in range(3)] for _ in range(3)]
        valid = False
        spaces = []

        def dfs(index):
            nonlocal valid
            if index == len(spaces):
                valid = True
                return

            i, j = spaces[index]
            for digit in range(9):
                if not line[i][digit] and not column[j][digit] and not block[
                        i // 3][j // 3][digit]:
                    line[i][digit] = True
                    column[j][digit] = True
                    block[i // 3][j // 3][digit] = True
                    board[i][j] = str(digit + 1)
                    dfs(index + 1)
                    line[i][digit] = False
                    column[j][digit] = False
                    block[i // 3][j // 3][digit] = False

                if valid:
                    return

        for i in range(9):
            for j in range(9):
                if board[i][j] == '.':
                    spaces.append((i, j))
                else:
                    digit = int(board[i][j]) - 1
                    line[i][digit] = True
                    column[j][digit] = True
                    block[i // 3][j // 3][digit] = True

        dfs(0)


# @lc code=end

File: leetcode-cn/C++/37.解数独.cpp
/*
 * @lc app=leetcode.cn id=37 lang=cpp
 *
 * [37] 解数独
 */

// @lc code=start
#include <string.h>
#include <vector>
using namespace std;

class Solution
{
public:
    bool hang[9][9];
    bool lie[9][9];
    bool block[9][9];
    vector<pair<int, int>> pos;
    bool ret = false;

    void help(int index, vector<vector<char>> &board)
    {
        if (index == pos.size())
        {
            ret = true;
            return;
        }

        auto [i, j] = pos[index];
        for (int num = 0; num < 9; num++)
        {
            if (!hang[i][num] && !lie[j][num] && !block[(i / 3) * 3 + j / 3][num])
            {
                if (ret)
                {
                    return;
                }

                board[i][j] = '1' + num;
                hang[i][num] = lie[j][num] = block[(i / 3) * 3 + j / 3][num] = true;
                help(index + 1, board);
                hang[i][num] = lie[j][num] = block[(i / 3) * 3 + j / 3][num] = false;
            }
        }
    }

    void solveSudoku(vector<vector<char>> &board)
    {
        memset(hang, false, sizeof(hang));
        memset(lie, false, sizeof(lie));
        memset(block, false, sizeof(block));

        for (int i = 0; i < 9; i++)
        {
            for (int j = 0; j < 9; j++)
            {
                if (board[i][j] == '.')
                {
                    pos.push_back({i, j});
                }
                else
                {
                    int num = board[i][j] - '1';
                    hang[i][num] = true;
                    lie[j][num] = true;
                    block[(i / 3) * 3 + j / 3][num] = true;
                }
            }
        }

        help(0, board);
    }
};
// @lc code=end
